tests/framework/extensions/charts/ranchergatekeeper.go

```diff
@@ -0,0 +1,286 @@
+	Resource: "customresourcedefinitions",
+}
+
+// InstallRancherGatekeeperChart installs the OPA gatekeeper chart
+func InstallRancherGatekeeperChart(client *rancher.Client, installOptions *InstallOptions) error {
+	hostWithProtocol := fmt.Sprintf("https:// %s", client.RancherConfig.Host)
+	gatekeeperChartInstallActionPayload := &payloadOpts{
+		InstallOptions: *installOptions,
+		Name:           RancherGatekeeperName,
+		Host:           hostWithProtocol,
+		Namespace:      RancherGatekeeperNamespace,
+	}
+
+	chartInstallAction := newGatekeeperChartInstallAction(gatekeeperChartInstallActionPayload)
+
+	catalogClient, err := client.GetClusterCatalogClient(installOptions.ClusterID)
+	if err != nil {
+		return err
+	}
+
+	// Cleanup registration
+
+	// register uninstall rancher-gatekeeper as a cleanup function
+	client.Session.RegisterCleanupFunc(func() error {
+		// UninstallAction for when uninstalling the rancher-gatekeeper chart
+		defaultChartUninstallAction := newChartUninstallAction()
+
+		err := catalogClient.UninstallChart(RancherGatekeeperName, RancherGatekeeperNamespace, defaultChartUninstallAction)
+		if err != nil {
+			return err
+		}
+
+		watchAppInterface, err := catalogClient.Apps(RancherGatekeeperNamespace).Watch(context.TODO(), metav1.ListOptions{
+			FieldSelector:  "metadata.name=" + RancherGatekeeperName,
+			TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = wait.WatchWait(watchAppInterface, func(event watch.Event) (ready bool, err error) {
+			if event.Type == watch.Error {
+				return false, fmt.Errorf("there was an error uninstalling rancher gatekeeper chart")
+			} else if event.Type == watch.Deleted {
+				return true, nil
+			}
+			return false, nil
+		})
+		if err != nil {
+			return err
+		}
+
+		dynamicClient, err := client.GetDownStreamClusterClient(installOptions.ClusterID)
+		if err != nil {
+			return err
+		}
+
+		namespaceResource := dynamicClient.Resource(namespaces.NamespaceGroupVersionResource).Namespace("")
+
+		err = namespaceResource.Delete(context.TODO(), RancherGatekeeperNamespace, metav1.DeleteOptions{})
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+
+		customResourceDefinitionResource := dynamicClient.Resource(CustomResourceDefinition).Namespace("")
+		CRDs, err := customResourceDefinitionResource.List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			return err
+		}
+
+		for _, unstructuredCRDs := range CRDs.Items {
+			newCRD := &apiextensionsv1.CustomResourceDefinition{}
+			err := scheme.Scheme.Convert(&unstructuredCRDs, newCRD, unstructuredCRDs.GroupVersionKind())
+			if err != nil {
+				return err
+			}
+
+			CRDname := newCRD.Spec.Names.Plural + "." + newCRD.Spec.Group
+
+			if strings.Contains(CRDname, "gatekeeper") {
+				err := customResourceDefinitionResource.Delete(context.TODO(), CRDname, metav1.DeleteOptions{})
+				if err != nil {
+					return err
+				}
+
+			}
+
+		}
```

Comment:
Yes completely agree.

Reply:
Just to clarify, you're saying separate this block out into three functions, and put them in a package like `tests/framework/extensions/customresourcedefinitions.go` ?

Reply:
Yes Natalie, partly correct, separating this block into three functions would be great for reusability. 

The naming is mostly personal preferences, so please read below as an example instead of a todo, you can do something like this:

- Put them in a package called `customresourcedefinitions`, and folder `tests/framework/extensions/customresourcedefinitions`.

And then for each helper, my way to do this would be to put them into their files as:

- List - `tests/framework/extensions/customresourcedefinitions/list.go`
- Delete - `tests/framework/extensions/customresourcedefinitions/delete.go`
- A string check helper - `tests/framework/extensions/customresourcedefinitions/customresourcedefinitions.go`